Lib/test/_test_multiprocessing.py

```diff
@@ -4680,6 +4682,150 @@ def test__all__(self):
         # Just make sure names in blacklist are excluded
         support.check__all__(self, multiprocessing, extra=multiprocessing.__all__,
                              blacklist=['SUBDEBUG', 'SUBWARNING'])
+
+
+class SpyReducer(reduction.AbstractReducer):
+    """Custom reducer that records the call to the Pickler
+    class it returns. Mocks are not used because they do not
+    play good with pickling"""
+    def __init__(self, spy):
+        self.spy = spy
+    def get_pickler(self):
+        spy = self.spy
+        class Pickler(pickle.Pickler):
+            def __init__(self):
+                pass
+            @classmethod
+            def dumps(self, obj, protocol=None):
+                spy["dumps"] += 1
+                return pickle.dumps(obj)
+            @classmethod
+            def loads(self, *args, **kwargs):
+                spy["loads"] += 1
+                return pickle.loads(*args, **kwargs)
+        return Pickler()
+
+class _TestCustomReducer(BaseTestCase):
+
+    ALLOWED_TYPES = ('processes',)
+
+    def setUp(self):
+        self.context = multiprocessing.get_context()
+        self.original_reducer = multiprocessing.get_reducer()
+        self.spy = {'loads': 0, 'dumps': 0}
+        self.context.set_reducer(SpyReducer(self.spy))
+
+    def tearDown(self):
+        self.context.set_reducer(self.original_reducer)
+        del self.spy
+
+    @unittest.skipUnless(HAS_REDUCTION, "test needs multiprocessing.reduction")
+    def test_connection_custom_reducer(self):
+        with self.connection.Listener() as l:
+            with self.connection.Client(l.address) as c:
+                with l.accept() as d:
+                    c.send(1729)
+                    self.assertEqual(d.recv(), 1729)
+
+        if self.TYPE == 'processes':
+            self.assertRaises(OSError, l.accept)
+
+        self.assertEqual(self.spy['loads'], 1)
+        self.assertEqual(self.spy['dumps'], 1)
+
+    @classmethod
+    def _put_and_get_in_queue(cls, queue, parent_can_continue):
+        parent_can_continue.set()
+        queue.put("Something")
+        queue.get()
+
+    @unittest.skipUnless(HAS_REDUCTION, "test needs multiprocessing.reduction")
+    def test_queue_custom_reducer(self):
+        queue = self.Queue()
+        parent_can_continue = self.Event()
+        p = self.Process(target=self._put_and_get_in_queue,
+            args=(queue, parent_can_continue))
+        p.daemon = True
+        p.start()
+        parent_can_continue.wait()
+        element = queue.get(timeout=TIMEOUT3)
+        self.assertEqual(element, "Something")
+        queue.put("Other_Something")
+        p.join(timeout=TIMEOUT3)
+        self.assertEqual(p.exitcode, 0)
+        self.assertEqual(element, "Something")
+        self.assertEqual(self.spy['loads'], 1)
+        self.assertEqual(self.spy['dumps'], 1)
+        close_queue(queue)
+
+
+class CustomContext(multiprocessing.context.BaseContext):
```

Comment:
Hmm... I think you misunderstood me. By "custom context" it did not mean add a custom Context class. I meant have one test class which sets the default reducer (`mp.set_reducer(...)`) and one test class which sets the reducer on a non-default context (`mp.get_context("some method").set_reducer(...)`).

Reply:
Do you mean so we do the test over the non default context and check that the default reducer was not called or the other way around? (Or something completely different?)

Reply:
> Do you mean so we do the test over the non default context and check that the default reducer was not called or the other way around?

I would say:
* test over the non-default context and check the non-default reducer was called
* test over the default context and check the non-default reducer was _not_ called